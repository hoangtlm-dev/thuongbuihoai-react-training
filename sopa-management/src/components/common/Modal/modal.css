.cart {
  width: 605px;
  padding: 35px 40px 0;
}

.cart-header,
.cart-item,
.cart-footer {
  display: flex;
  justify-content: space-between;
}

.cart-image-item {
  width: 130px;
  height: 130px;
  border-radius: 24px;
  border: 1px solid #E3E5ED;
  background: #FFF;
  margin-right: 20px;
}

.card-body {
  margin: 110px 0 428px;
}

.cart-item {
  margin-bottom: 50px;
}

.cart-info {
  display: flex;
}

.card-quantity {
  display: flex;
}

.cart-color {
  margin: 16px 0 12px;
}

.cart-size {
  margin-bottom: 24px;
}

.cart-btn {
  margin-left: 16px;
}

.cart-footer {
  margin-top: 24px;
}

.cart-text {
  font-weight: 600;
  font-size: var(--text-base);
  line-height: var(--leading-primary);
}