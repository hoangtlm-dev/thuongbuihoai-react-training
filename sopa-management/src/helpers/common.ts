// Flattens an array of arrays into a single array
export const flattenArray = <T>(pages: T[][]): T[] => {
  const result: T[] = [];
  pages.forEach((page) => {
    result.push(...page);
  });

  return result;
};
