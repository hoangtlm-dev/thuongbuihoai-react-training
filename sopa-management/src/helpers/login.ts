import {
  Account,
  LoginType
} from "@interfaces/account";
import { FieldError } from "react-hook-form";

export enum CheckType {
  login = 'login',
  email = 'email',
  password = 'password'
}

// Check login
export const checkAccount = (
  data?: Account[],
  email?: string,
  password?: string,
  checkType?: CheckType,
) => {
  switch (checkType) {
    case CheckType.login:
      return data?.some((user: Account) => user.email === email && user.password === password);
    case CheckType.email:
      return data?.some((user: Account) => user.email === email);
    case CheckType.password:
      return data?.some((user: Account) => user.password === password);
    default:
      throw new Error('Invalid check type');
  }
};

export const checkValidationStyles = (
  serverError: boolean,
  rules: FieldError | undefined,
  isDirty: boolean,
  type: LoginType,
) => {
  const checkIsDirty = isDirty ? type.info : type.default
  const style = serverError || rules ? type.error : checkIsDirty;

  return { style };
};