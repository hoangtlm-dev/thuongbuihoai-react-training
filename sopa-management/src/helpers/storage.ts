import { Product } from "../interfaces/product";

export const setStorage = (key: string, data: Product) => {
  localStorage.setItem(key, JSON.stringify(data));
};

export const getStorage = (key: string) => {
  return JSON.parse(localStorage.getItem(key) as any) || [];
};

export const clearStorage = () => {
  localStorage.clear();
};
