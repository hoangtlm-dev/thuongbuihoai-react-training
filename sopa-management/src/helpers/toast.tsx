import Toast from "../components/Toast";

export const showToast = (
  title: string,
  message: string,
  type: 'success' | 'error',
  duration: number
) => {
  return (
    <Toast
      title={title}
      message={message}
      type={type}
      duration={duration}
    />
  )
};
