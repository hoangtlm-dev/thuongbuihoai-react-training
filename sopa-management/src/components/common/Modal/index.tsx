import { memo } from "react";
import './modal.css';
import { Trash } from "../../../../public/images/icons";
import Image from "../Image";
import Text, { ThemeType } from "../Text";
import Counter from "../Counter";
import Button, { ButtonType } from "../Button";
import Price, { PriceType } from "../Price";
import { Product } from "../../../interfaces/product";
import { useCounterStore } from "../../../stores/counter";
import { getStorage } from "../../../helpers/storage";
import { totalPrices } from "../../../helpers/common";

const Modal: React.FC = () => {
  const { count, setCount } = useCounterStore();
  const cartData = getStorage('cart') || [];

  const handleIncrement = () => {
    setCount(count + 1);
  };

  const handleDecrement = () => {
    setCount(count - 1);
  };


  return (
    <div className="cart">
      <div className="cart-header">
        <Text text="Cart" className="cart-text" />
        <Trash />
      </div>
      <div className="card-body">
        {cartData?.map((item: Product) => (
          <div className="cart-item">
            <div className="cart-info">
              <figure className="cart-image">
                <Image className="cart-image-item" url={item.image.url} />
              </figure>
              <div className="cart-description">
                <Text text={item.name} className="cart-text" />
                <Text text={`Color: ${item.color}`} type={ThemeType.highlightPrimary} className="cart-color" />
                <Text text={`Size: ${item.size}`} type={ThemeType.highlightPrimary} className="cart-size" />
                <div className="card-quantity">
                  <Counter
                    value={item?.quantity}
                    onDecrement={handleDecrement}
                    onIncrement={handleIncrement}
                  />
                  <Button children={<Trash />} type={ButtonType.btnIconSecondary} className="cart-btn" />
                </div>
              </div>
            </div>
            <Price value={item.price} type={PriceType.tertiary} className="cart-price" />
          </div>
        ))}
      </div>
      <hr />
      <div className="cart-footer">
        <Text text="Subtotal" className="cart-text" />
        <Price value={totalPrices(cartData)} type={PriceType.tertiary} />
      </div>
    </div>
  )
}

export default memo(Modal);
