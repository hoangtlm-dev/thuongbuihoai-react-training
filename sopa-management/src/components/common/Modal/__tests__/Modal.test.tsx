// Libs
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

// Helpers
import { renderRouterTest } from '../../../../helpers/testUtils';

// Components
import Modal from '../index';

describe('Modal component', () => {
  test('Should render Modal component by default', () => {
    const { getByTestId } = renderRouterTest(<Modal />);
    const modal = getByTestId('modal');
    expect(modal).toBeInTheDocument();
  });

  test('Should render Modal component by click close button', async () => {
    const onClickCloseButton = jest.fn();
    const { getByTestId } = renderRouterTest(
      <Modal onToggleModal={onClickCloseButton} />
    );
    const modal = getByTestId('button');
    await userEvent.click(modal);
    expect(onClickCloseButton).toHaveBeenCalled();
  });

  test('Matches snapshot', () => {
    const { container } = renderRouterTest(<Modal />);
    expect(container).toMatchSnapshot();
  });
});
