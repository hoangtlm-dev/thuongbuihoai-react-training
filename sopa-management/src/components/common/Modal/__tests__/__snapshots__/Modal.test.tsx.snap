// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component Matches snapshot 1`] = `
<div>
  <div
    class="overlay"
    data-testid="modal"
  >
    <div
      class="cart"
    >
      <div
        class="cart-header"
      >
        <p
          class="cart-text text-"
          data-testid="text"
        >
          Cart
        </p>
        <button
          class=" btn btn-icon-primary"
          data-testid="button"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6L6 18"
              stroke="#02021D"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M6 6L18 18"
              stroke="#02021D"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="card-body"
      />
      <hr />
      <div
        class="cart-footer"
      >
        <p
          class="cart-text text-"
          data-testid="text"
        >
          Subtotal
        </p>
        <p
          class=" price-tertiary"
          data-testid="price"
        >
          $
          0
        </p>
      </div>
    </div>
  </div>
</div>
`;
